### PR TITLE
withdraw, swap and deposit from 3pool to 3eur + stake into convex

### PR DESCRIPTION
**Pull request by jorijnsmit**
if encountering errors, replace `Contract(` with `Contract.from_explorer(` in curve.py on the first run.
```
snapshot result for 0xD0A7A8B98957b9CD3cFB9c0425AbE44551158e9e:
                    balance_before             balance_after                balance_delta
symbol                                                                                   
3Crv    671,048.418387414489367007      0.000000000000000000  -671,048.418387414489367007
USDC          0.000000000000000000  1,369.370748000000000000     1,369.370748000000000000
EURS          0.000000000000000000  1,179.080000000000000000     1,179.080000000000000000 
```

### CODE REVIEW

File: helpers/addresses.py [hunk 1]
```diff
@@ -146,6 +146,8 @@
         "yvWBTC": "0xA696a63cc78DfFa1a63E9E50587C197387FF6C7E",  ##TODO NO COINGECKO PRICE
         "aBADGER": "0x43298F9f91a4545dF64748e78a2c777c580573d6",
         "badgerWBTC_f": "0x137469B55D1f15651BA46A89D0588e97dD0B6562",
+        "EURS": "0xdB25f211AB05b1c97D595516F45794528a807ad8",
+        "crv3eur": "0xb9446c4Ef5EBE66268dA6700D26f96273DE3d571",
```
Comment (sajanrajdev) [inline helpers/addresses.py:150]: Confirmed both addresses

**Comment by sajanrajdev** [comment]
I ran the curve swap tests and got the stable swap test to fail with the following error:
image

Did you not get this?

**Comment by jorijnsmit** [comment]
yea so looks like usdc is already introduced to your db with interface.erc20, which doesnt have the mint method.

if you uncomment the line above that, and thus force a Contract.from_explorer on the usdc address, it solves it.

i was looking at `MintableForkToken` as well, and maybe create an issue in brownie-token-tester. i believe it needs its own manual interface (like an extended erc20, with `mint`, `createTokens`, `configureMinter` etc instead of just inheriting from `Contract`.

File: helpers/addresses.py (hunk 1 above)
Comment (petrovska-petro) [reply helpers/addresses.py:150]: +1

**Comment by sajanrajdev** [comment]
> yea so looks like usdc is already introduced to your db with interface.erc20, which doesnt have the mint method.
> 
> if you uncomment the line above that, and thus force a Contract.from_explorer on the usdc address, it solves it.
> 
> i was looking at `MintableForkToken` as well, and was thinking to maybe create an issue in brownie-token-tester. i believe it needs its own manual interface (like an extended erc20, with `mint`, `createTokens`, `configureMinter` etc) instead of just inheriting from `Contract`.

What's the advantage of this over `distribute_from_whales` from badger-utils?

**Comment by jorijnsmit** [comment]
> What's the advantage of this over `distribute_from_whales` from badger-utils?

@sajanrajdev: mainly the fact that brownie-token-test most of the times is a mint instead of a transfer, which is a bit cleaner. what i remember from comparing is that the whale distribute had quite a few onchain calls as overhead, making it tough to separate the script from the whale distribution.

**Review by sajanrajdev — DISMISSED** [review]
Script moves safe's balance of 3pool into the 3Eur Curve pool and takes it on Curve. The script verifies the USD/EUR conversion and also checks the correct staking on fork. LGTM. Snapshot:
```
snapshot result for 0xD0A7A8B98957b9CD3cFB9c0425AbE44551158e9e:
                    balance_before             balance_after                balance_delta
symbol
3Crv    671,048.418387414489367007      0.000000000000000000  -671,048.418387414489367007
USDC          0.000000000000000000  1,369.369964000000000000     1,369.369964000000000000
EURS          0.000000000000000000  1,183.190000000000000000     1,183.190000000000000000
```

Changes to the tests LGTM and are passing:
image

File: great_ape_safe/ape_api/curve.py [hunk 2]
```diff
@@ -66,36 +76,31 @@ def _get_coin_indices(self, lp_token, pool, asset_in, asset_out):
...
-            return registry.get_n_coins(pool)[0] # note [1] tells us if there is a wrapped coin!
+        return registry.get_n_coins(pool)[0]
+
+
+    def _pool_has_wrapped_coins(self, pool):
+        registry = self._get_registry(pool)
+        try:
+            registry.get_underlying_balances(pool)
+            return True
+        except VirtualMachineError:
```
Comment (petrovska-petro) [inline great_ape_safe/ape_api/curve.py:101]: for me locally here seems to jump instead a `AttributeError`, do you get other type of exception?

**Review by petrovska-petro — APPROVED** [review]
Looking good! 

Snap, usdc and eurs delta from `DUSTY` constant :

```
snapshot result for 0xD0A7A8B98957b9CD3cFB9c0425AbE44551158e9e:
                    balance_before             balance_after                balance_delta
symbol                                                                                   
3Crv    671,048.418387414489367007      0.000000000000000000  -671,048.418387414489367007
USDC          0.000000000000000000  1,369.304557000000000000     1,369.304557000000000000
EURS          0.000000000000000000  1,186.380000000000000000     1,186.380000000000000000 
```

**Review by jalbrekt85 — APPROVED** [review]
confirmed that `curve.swap` now works using registry exchange. appropriate amount of funds are withdrawn and staked in convex.

**Pull request merged**